libs/dokobit-signing/src/lib/signing.service.ts

```diff
@@ -0,0 +1,159 @@
+import { Base64 } from 'js-base64'
+import { createHash } from 'crypto'
+
+import { Injectable, Inject } from '@nestjs/common'
+
+import { Logger, LOGGER_PROVIDER } from '@island.is/logging'
+
+export interface SigningServiceOptions {
+  url: string
+  accessToken: string
+}
+
+export class SigningServiceResponse {
+  controlCode: string
+  documentToken: string
+}
+
+export class SigningServiceError extends Error {
+  constructor(code: number, message: string) {
+    super()
+
+    this.name = 'DokobitError'
+    this.message = `${code}: ${message}`
+  }
+}
+
+interface DokobitSignResponse {
+  status: string
+  control_code: string
+  token: string
+  error_code: number
+  message: string
+}
+
+interface DokobitStatusResponse {
+  status: string
+  file: {
+    name: string
+    digest: string
+    content: string
+  }
+  signature_id: string
+  error_code: number
+  message: string
+}
+
+@Injectable()
+export class SigningService extends RESTDataSource {
+  constructor(
+    @Inject('SIGNING_OPTIONS') private options: SigningServiceOptions,
+    @Inject(LOGGER_PROVIDER)
+    private logger: Logger,
+  ) {
+    super()
+
+    this.baseURL = `${this.options.url}/mobile/`
+  }
+
+  willSendRequest(request: RequestOptions) {
+    request.headers.set('Content-Type', 'multipart/form-data')
+  }
+
+  async requestSignature(
+    mobileNumber: string,
+    message: string,
+    contact: string,
+    location: string,
+    documentName: string,
+    documentContent: string,
+  ): Promise<SigningServiceResponse> {
+    this.logger.debug(
+      `${contact} ${location} is requesting ${mobileNumber} to sign ${documentName} with message ${message}`,
+    )
+
+    const base64 = Base64.encode(documentContent)
+    const digest = createHash('sha256')
+      .update(base64)
+      .digest('hex')
+
+    const body = new FormData()
+    body.append('phone', mobileNumber)
+    body.append('message', `${message} `)
+    body.append('timestamp', 'true')
+    body.append('language', 'IS')
+    body.append('pdf[contact]', contact)
+    body.append('pdf[location]', location)
+    body.append('type', 'pdf')
+    body.append('pdf[files][0][name]', documentName)
+    body.append('pdf[files][0][content]', base64)
+    body.append('pdf[files][0][digest]', digest)
+
+    const resSign: DokobitSignResponse = await this.post(
+      `sign.json?access_token=${this.options.accessToken}`,
+      body,
+    )
+
+    if (resSign.status !== 'ok') {
+      throw new SigningServiceError(resSign.error_code, resSign.message)
+    }
+
+    return {
+      controlCode: resSign.control_code,
+      documentToken: resSign.token,
+    }
+  }
+
+  async getDocument(
+    documentName: string,
+    documentToken: string,
+  ): Promise<string> {
+    let resStatus: DokobitStatusResponse
+
+    // Try to retrieve the signed document
+    // Need to try longer than the mobile signature timeout, but not too long
+    for (let i = 1; i < 120; i++) {
```

Comment:
This seems like a code smell. Why do we need this  if we are awaiting `this.post`

Reply:
I agree, but this is how the Dokobit API works and how the code examples from Dokobit do this.

The API returns pretty much immediately, whether the document has been signed or not. At the same time, the user gets a long time to respond to the signature request on the mobile phone.

The result is that we need to try repeatedly until we get an answer other than ´waiting´. I don't know of a good way to do this differently.